Make resource drop tooltip consistent between 2D and 3D

Nodragem opened the pull request:
Makes messages and tooltips consistent when dropping Mesh, Scene, Texture, Material in the 2D or 3D Viewport. There were no message in 3D Viewport when dropping Mesh/Scene before.
Closes: https://github.com/godotengine/godot-proposals/issues/9752

As the PR evolved though the years, I also modernised the tooltips so it looks better (i.e. transparent panel, theme accent color, shortcuts shown in bold).

I would also like to add consistency for AudioStream but need to coordinate with: #92004

Demo:

https://github.com/user-attachments/assets/d3e8aa46-8778-4728-8561-34a0f6150303

Test project:
[testing_assets_4.3.zip](https://github.com/godotengine/godot/files/15365543/testing_assets_4.3.zip)

Nodragem committed:
adding tooltips when dropping files

AThousandShips commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -5741,17 +5741,20 @@ void CanvasItemEditorViewport::_create_preview(const Vector<String> &files) cons
-						label_desc->set_text(desc);
+						dropping_tooltip_label->show();
+						dropping_tooltip_label_desc->show();
+						dropping_tooltip_label->set_text(vformat(TTR("Adding Scene...")));
```
```suggestion
						dropping_tooltip_label->set_text(TTR("Adding Scene..."));
```

Nodragem replied:
I was'nt sure why there was a `vformat` in some of the labels; what is it for?

AThousandShips replied:
It's to format, replaces things like `%s`

Nodragem commented:
I need some help.

Let's start with what is working.
I resolved the merge conflicts I had with @timothyqiu changes by favouring their new way of organising the code. 
For instance:
- I kept the `set_text()/show()` tooltips in the `can_drop_data()` function for the 2D viewport as @timothyqiu did
- I moved of the `set_text()/show()` tooltips in the `can_drop_data_fw()` function for the 3D viewport to be consistent with @timothyqiu  
- I also introduced the `instantiate_type` enum variable in the 2D code, again for being consistent

I changed the text of the tooltips so that they work with mesh, audio, scene, texture, materials.

The 2D drag-and-drop seems to work well.
https://github.com/user-attachments/assets/6ac1d6df-9a00-45d3-b6b9-f1af5a61d2be

However, on the 3D side, I have the problem that the tooltips appear from 1 frame and then they disappear!
As shown below:
https://github.com/user-attachments/assets/f6c3cc47-d4df-4ebc-9978-642bf3322c0e

I set up the C++ debugger in VS Code to be able to check if there was any `dropping_tooltip_label->hide()` being called unexpectedly, but no luck. I just can't find the bug!

Nodragem commented:
Some progress in my bug hunting.

 I found that `can_drop_data_fw()` seems to be called every frame 😅 hehe! and that's is causing the issue. On the second time, the preview node is in the tree and under this condition, I end up with an empty string in my tooltip.

So basically, I need prevent my tooltip to get erased when can_drop_data_fw() is called and a preview node is in the tree.

Nodragem commented:
Should be fixed now.

timothyqiu reviewed (dismissed):
* Dropping a Texture in the 2D editor is missing the messages about holding Shift and holding Alt.
* When dropping multiple types of file, the messages about modifier keys are also missing.
* There're cases where the message says "Dropping a A, B, or C". It feels wrong because we can definitely tell whether it's A, B, or C in code.

Nodragem commented:
> * Dropping a Texture in the 2D editor is missing the messages about holding Shift and holding Alt.
> 
>     * When dropping multiple types of file, the messages about modifier keys are also missing.
> 
>     * There're cases where the message says "Dropping a A, B, or C". It feels wrong because we can definitely tell whether it's A, B, or C in code.

- I will have a look concerning 1.
- Concerning 2, when there are multiple files, I chose to not show any messages for now, but I will have a look at what the current behaviour of the modifier keys is.
- Concerning 3, the idea was that for A, B, C the modifier keys do the same. I will see if I can make it as you suggested.

Nodragem commented:
@timothyqiu, I should have covered all your requests now.

I also made it so that there is a warning message when the File is not supported. At the moment it reads like this: `File format is not supported... Dropping Unrecognized File(s)`. Happy to modify this message to whatever the reviewers think is best.

In terms of variable names, this PR has given the names `dropping_tooltip_label` and `dropping_tooltip_label_desc` to the labels used by the drag-and-drop tooltip. I am considering to rename them to `tooltip_title` and `tooltip_desc` as I believe that these labels could be used by other tools? Happy to hear any other suggestions.

Calinou commented on editor/plugins/node_3d_editor_plugin.cpp:
```diff
@@ -5723,23 +5775,20 @@ Node3DEditorViewport::Node3DEditorViewport(Node3DEditor *p_spatial_editor, int p
+	tooltip_label->hide();
+	vbox->add_child(tooltip_label);
+
+	//TODO: MacOS ALT/CTRL changed to Cmd
```
This comment can be removed now:

```suggestion
```

Nodragem replied:
Well spotted, just pushed an update!

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/45875c2a6cc74a01596761aba21515fa756c7faa), it works as expected.

### 2D

![image](https://github.com/user-attachments/assets/3f74d219-d3d4-498a-b1b8-6667217e2e8d)

![image](https://github.com/user-attachments/assets/9cdf29c5-9571-4e96-8584-94d1dafe70bd)

![image](https://github.com/user-attachments/assets/d05604dc-cc10-4753-8dac-f5b355eddb7a)


### 3D

![image](https://github.com/user-attachments/assets/fc5ca116-aace-43b7-9a62-997f90f42d6b)

![image](https://github.com/user-attachments/assets/2c5ef476-989f-4235-9374-5800319bc001)

![image](https://github.com/user-attachments/assets/ae2386be-3fdf-4310-8c90-ab7ea792cf5f)

There are some stylistic changes we could make (such as using a font outline instead of a shadow as it's more readable), but this can be left to a future PR. We should harmonize this throughout the editor to use a style similar to the 3D editor ruler text, or the freelook zoom/speed text in the indicator at the left of the 3D edcitor viewport.

Nodragem commented:
> There are some stylistic changes we could make (such as using a font outline instead of a shadow as it's more readable), but this can be left to a future PR. We should harmonize this throughout the editor to use a style similar to the 3D editor ruler text, or the freelook zoom/speed text in the indicator at the left of the 3D edcitor viewport.

Totally agree! but I was more thinking about using a semi-transparent card/panel such as being use for on-screen information:
![image](https://github.com/user-attachments/assets/a235fc0d-3e10-4477-a435-c442d43b3aaa)

And maybe using the editor `Accent Color` which is use for title in tooltips and documentation instead of the yellow for the header title, I don't know?
![image](https://github.com/user-attachments/assets/0653635f-8c9d-455e-a3d8-fbfa52584d23)
![image](https://github.com/user-attachments/assets/09c35825-931d-4355-98bb-bbdbe7043d35)

AThousandShips commented:
The rebase reset the author of those commits, probably because you pushed without `--force` which created a merge commit, you can resolve that by resetting your branch and trying again

Nodragem commented on editor/scene/3d/node_3d_editor_plugin.cpp:
```diff
@@ -4965,18 +4991,44 @@ bool Node3DEditorViewport::can_drop_data_fw(const Point2 &p_point, const Variant
+				"\n" + TTR("Hold Alt: added as child of root node.");
+
+		if (files.size() > 1) {
+			title = TTR("Dropping multiple Files ...");
```
even went for lower case F for consistency 😃

Nodragem committed:
adding panel and richlabeltext

Nodragem committed:
improve tooltips when dropping files in viewport

remove can_instantiate

kitbdev reviewed (commented):
> I did not squash the commits yet, because I am not sure how to do it without deleting my own history. I want to keep a version where I used a Panel+Label+RichTextLabel in case this new version with just a RichTextLabel is not good enough.

If you want to squash this PR, you can make a new backup branch locally to keep the original history.

AThousandShips commented on editor/scene/3d/node_3d_editor_plugin.cpp:
```diff
@@ -5521,93 +5400,126 @@ void Node3DEditorViewport::_perform_drop_data() {
 	}
 }
 
+void Node3DEditorViewport::_show_tooltip(String title, String description) const {
```
```suggestion
void Node3DEditorViewport::_show_tooltip(const String &p_title, const String &p_description) const {
```

Nodragem commented on editor/scene/3d/node_3d_editor_plugin.cpp:
```diff
@@ -5618,17 +5530,43 @@ bool Node3DEditorViewport::can_drop_data_fw(const Point2 &p_point, const Variant
+
+	desc = "[ul]";
+	if (instantiate_type != 0) {
+		desc += TTR("[b]Default:[/b] Added as sibling of selected node (except when root is selected).") +
```
“Added” is not past tense here, it is the past participle, which acts as an adjective for the file being dropped, so it does not strike me as being inconsistent with the future tense (you can say “it will be added”).

But yes, maybe better to have it phrased as an action, as it is done in menus. I would keep the “as” though. Maybe something like `Instantiate as sibling of selected nodes` ?

AdriaandeJongh reviewed (approved):
Didn't review the code, but tested the PR and it looks good overall.

I'm a bit bothered by the big amount of horizontal spacing here, before the bullets. I suspect if this PR doesn't address it, nobody will 😅 Is there an easy way to reduce that spacing there? Perhaps even stop using the list and insert the bullet point characters at the start manually?

<img width="629" height="234" alt="Screenshot 2026-03-05 at 14 26 33" src="https://github.com/user-attachments/assets/3a437b1b-931c-4f92-b2a7-f3be2bda4668" />

kitbdev commented on editor/scene/3d/node_3d_editor_plugin.cpp:
```diff
@@ -5649,17 +5530,47 @@ bool Node3DEditorViewport::can_drop_data_fw(const Point2 &p_point, const Variant
+	}
+
+	String title = "Dropping unsupported file(s)";
+	String desc = "File format is not supported...";
```
Should probably be translated? Though I'm not sure this will ever show if its filtered before it gets here I guess.

Nodragem commented:
> I'm a bit bothered by the big amount of horizontal spacing here, before the bullets. I suspect if this PR doesn't address it, nobody will 😅 Is there an easy way to reduce that spacing there? Perhaps even stop using the list and insert the bullet point characters at the start manually?

I wonder if we can use the `tab_size` parameter as hinted there: https://docs.godotengine.org/en/stable/classes/class_richtextlabel.html#class-richtextlabel-method-push-list

kitbdev commented on editor/scene/3d/node_3d_editor_plugin.cpp:
```diff
@@ -5649,17 +5530,47 @@ bool Node3DEditorViewport::can_drop_data_fw(const Point2 &p_point, const Variant
+
+	_show_tooltip(title, desc);
 
 	return false;
```
This should still return true if it is a valid drop, probably the same conditions that can return true above.
Otherwise it is possible to have the cannot drop mouse cursor yet show that it can on the info tooltip, if there is only one mouse move event. It also may cause issues if someone tries to drag and drop with scripted mouse events.

Nodragem replied:
I don't think I modified this part of the code, it was already returning false. I changed it to true now, and nothing has broken (may need some testing).

AThousandShips commented on editor/scene/canvas_item_editor_plugin.h:
```diff
@@ -32,6 +32,7 @@
 
 #include "editor/plugins/editor_plugin.h"
 #include "scene/gui/box_container.h"
+#include "scene/gui/rich_text_label.h"
```
```suggestion
```
Same here

Nodragem commented:
@AdriaandeJongh here how it looks now:

<img height="180" alt="Screenshot 2026-03-06 110558" src="https://github.com/user-attachments/assets/f3c865be-637d-4a60-bf73-11ca6a834639" />

<img  height="180" alt="Screenshot 2026-03-06 110612" src="https://github.com/user-attachments/assets/8e58cf38-04d9-42c2-9961-c1ec080293a8" />

<img height="250" alt="Screenshot 2026-03-06 111903" src="https://github.com/user-attachments/assets/06852340-4818-4cad-949f-3b8d39f55c09" />

Note that I also change the title for unsupported format to "Can't drop the file...".

AdriaandeJongh reviewed (approved):
Tested it. The new addition of the error message also appearing in the tooltip looks great to me!

AThousandShips reviewed (approved):
Code style looks good, haven't tested
